nixos/tests/installer.nix

```diff
@@ -68,160 +67,180 @@ let
 
-      # FIXME don't duplicate the -enable-kvm etc. flags here yet again!
-      qemuFlags =
-        (if system == "x86_64-linux" then "-m 768 " else "-m 512 ") +
-        (optionalString (system == "x86_64-linux") "-cpu kvm64 ") +
-        (optionalString (system == "aarch64-linux") "-enable-kvm -machine virt,gic-version=host -cpu host ");
+      makeMachineConfig = name: pkgs.writeText "machine-config.json" (builtins.toJSON ({
+        inherit name;
+        # FIXME don't duplicate the -enable-kvm etc. flags here yet again!
+        qemuFlags =
+          (if system == "x86_64-linux" then "-m 768 " else "-m 512 ") +
+          (optionalString (system == "x86_64-linux") "-cpu kvm64 ") +
+          (optionalString (system == "aarch64-linux") "-enable-kvm -machine virt,gic-version=host -cpu host ");
+
+        hda = "vm-state-machine/machine.qcow2";
+        hdaInterface = iface;
+      } // (optionalAttrs isEfi {
+        bios = if pkgs.stdenv.isAarch64
+          then "${pkgs.OVMF.fd}/FV/QEMU_EFI.fd"
+          else "${pkgs.OVMF.fd}/FV/OVMF.fd";
+      })));
 
-      hdFlags = ''hda => "vm-state-machine/machine.qcow2", hdaInterface => "${iface}", ''
-        + optionalString isEfi (if pkgs.stdenv.isAarch64
-            then ''bios => "${pkgs.OVMF.fd}/FV/QEMU_EFI.fd", ''
-            else ''bios => "${pkgs.OVMF.fd}/FV/OVMF.fd", '');
     in if !isEfi && !(pkgs.stdenv.isi686 || pkgs.stdenv.isx86_64) then
       throw "Non-EFI boot methods are only supported on i686 / x86_64"
     else ''
+      def create_machine_from_file(machine_config_file):
+          with open(machine_config_file, "r") as f:
+              return create_machine(json.load(f))
+
 
-      $machine->start;
+      machine.start()
 
       # Make sure that we get a login prompt etc.
-      $machine->succeed("echo hello");
-      #$machine->waitForUnit('getty@tty2');
-      #$machine->waitForUnit("rogue");
-      $machine->waitForUnit("nixos-manual");
+      machine.succeed("echo hello")
+      # machine.wait_for_unit("getty@tty2")
+      # machine.wait_for_unit("rogue")
+      machine.wait_for_unit("nixos-manual")
 
       # Wait for hard disks to appear in /dev
-      $machine->succeed("udevadm settle");
+      machine.succeed("udevadm settle")
 
       # Partition the disk.
       ${createPartitions}
 
       # Create the NixOS configuration.
-      $machine->succeed("nixos-generate-config --root /mnt");
+      machine.succeed("nixos-generate-config --root /mnt")
 
-      $machine->succeed("cat /mnt/etc/nixos/hardware-configuration.nix >&2");
+      machine.succeed("cat /mnt/etc/nixos/hardware-configuration.nix >&2")
 
-      $machine->copyFileFromHost(
+      machine.copy_file_from_host(
           "${ makeConfig { inherit bootLoader grubVersion grubDevice grubIdentifier grubUseEfi extraConfig; } }",
-          "/mnt/etc/nixos/configuration.nix");
+          "/mnt/etc/nixos/configuration.nix",
```

Comment:
wouldn't this path be visible in the VM anyway as it's from the nix store path? wouldn't then `machine.succeed("cp ${makeConfig ...} /mnt/etc/nixos/configuration.nix")` work or is this suggestion stupid becasue i am missing out anything really fundamental?

Reply:
i dont think it is available in the VMs nix-store.